near-sdk-macros/src/core_impl/abi/abi_generator.rs

```diff
@@ -126,6 +127,34 @@ impl ImplItemMethodInfo {
                         None
                     }
                 }
+                ReturnType::Type(_, ty) if is_handles_result && utils::type_is_result(ty) => {
+                    let ty = if let Some(ty) = utils::extract_ok_type(ty) {
+                        ty
+                    } else {
+                        return syn::Error::new_spanned(
+                            ty,
+                            "Function marked with #[handle_result] should have return type Result<T, E> (where E implements FunctionError).",
+                        )
+                        .into_compile_error();
+                    };
```

Comment:
feels like this is code duplication from the extraction from non-abi codegen, but probably fine for now

Reply:
Yep, I kind of feel like the bindgen macro needs a pre-processing step that would do all of these "type" checks (`#[handle_result]`, `#[callback_vec]` etc) and produce a well-formed model. In this case, it would something like this:

```rust
enum BindgenReturnType {
    Regular(Ident),
    HandleResult { ident: Ident, ok_type: Ident, err_type: Ident },
}

struct BindgenFunction {
    ...
    return_type: BindgenReturnType
}
```

So that we can just safely use these models in later steps.